evm/bridge-token-factory/contracts/BridgeTokenFactory.sol

```diff
@@ -0,0 +1,318 @@
+
+        if (ECDSA.recover(hashed, signatureData) != nearBridgeDerivedAddress) {
+            revert InvalidSignature();
+        }
+
+        require(!_isBridgeToken[_nearToEthToken[metadata.token]], "ERR_TOKEN_EXIST");
+
+        address bridgeTokenProxy = address(
+            new ERC1967Proxy(
+                tokenImplementationAddress,
+                abi.encodeWithSelector(
+                    BridgeToken.initialize.selector,
+                    metadata.name,
+                    metadata.symbol,
+                    metadata.decimals
+                )
+            )
+        );
+
+        emit SetMetadata(
+            bridgeTokenProxy,
+            metadata.token,
+            metadata.name,
+            metadata.symbol,
+            metadata.decimals
+        );
+
+        _isBridgeToken[address(bridgeTokenProxy)] = true;
+        _ethToNearToken[address(bridgeTokenProxy)] = metadata.token;
+        _nearToEthToken[metadata.token] = address(bridgeTokenProxy);
+
+        return bridgeTokenProxy;
+    }
+
+    function setMetadata(
+        string calldata token,
+        string calldata name,
+        string calldata symbol
+    ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        require(_isBridgeToken[_nearToEthToken[token]], "ERR_NOT_BRIDGE_TOKEN");
+
+        BridgeToken bridgeToken = BridgeToken(_nearToEthToken[token]);
+
+        bridgeToken.setMetadata(name, symbol, bridgeToken.decimals());
+
+        emit SetMetadata(
+            address(bridgeToken),
+            token,
+            name,
+            symbol,
+            bridgeToken.decimals()
+        );
+    }
+
+    function deposit(bytes calldata signatureData, BridgeDeposit calldata bridgeDeposit) external whenNotPaused(PAUSED_DEPOSIT) {
+        if (completedTransfers[bridgeDeposit.nonce]) {
+            revert NonceAlreadyUsed();
+        }
+
+        bytes memory borshEncoded = bytes.concat(
+            Borsh.encodeUint128(bridgeDeposit.nonce),
+            Borsh.encodeString(bridgeDeposit.token),
+            Borsh.encodeUint128(bridgeDeposit.amount),
+            bytes1(0x00), // variant 1 in rust enum
+            Borsh.encodeAddress(bridgeDeposit.recipient),
+            bridgeDeposit.relayer == address(0)  // None or Some(Address) in rust
+                ? bytes("\x00") 
+                : bytes.concat(bytes("\x01"), Borsh.encodeAddress(bridgeDeposit.relayer))
+        );
+        bytes32 hashed = keccak256(borshEncoded);
+
+        if (ECDSA.recover(hashed, signatureData) != nearBridgeDerivedAddress) {
+            revert InvalidSignature();
+        }
+
+        require(_isBridgeToken[_nearToEthToken[bridgeDeposit.token]], "ERR_NOT_BRIDGE_TOKEN");
+        BridgeToken(_nearToEthToken[bridgeDeposit.token]).mint(bridgeDeposit.recipient, bridgeDeposit.amount);
+
+        completedTransfers[bridgeDeposit.nonce] = true;
+
+        emit Deposit(bridgeDeposit.token, bridgeDeposit.amount, bridgeDeposit.recipient);
```

Comment:
We need to emit the nonce to claim the fee on Near side.
Also, we need to pass a `string claim_recipient/fee_recipient` to identify the fee recipient

Reply:
Added it to signature as discussed 616d939c21e3ae02cec87bc2b27caa61a68335f5